app/src/main/kotlin/moe/tarsin/coroutines/NamedMutex.kt

```diff
@@ -1,23 +1,46 @@
 package moe.tarsin.coroutines
 
-import arrow.core.memoize
-import kotlin.contracts.InvocationKind
-import kotlin.contracts.contract
+import androidx.collection.mutableScatterMapOf
+import io.ktor.utils.io.pool.DefaultPool
 import kotlinx.coroutines.sync.Mutex
 import kotlinx.coroutines.sync.withLock
 
-class NamedMutex<T> {
-    val innerMutexGetter = { _: T -> Mutex() }.memoize()
-}
+class MutexTracker(
+    val mutex: Mutex = Mutex(),
+    var count: Int = 0,
+)
 
-suspend inline fun <T, K> NamedMutex<K>.withLock(key: K, owner: Any? = null, action: () -> T): T {
-    contract {
-        callsInPlace(action, InvocationKind.EXACTLY_ONCE)
-    }
+val MutexTracker.free
+    get() = count == 0
 
-    val mutex = synchronized(innerMutexGetter) {
-        innerMutexGetter(key)
+operator fun MutexTracker.inc() = apply { count++ }
+operator fun MutexTracker.dec() = apply { count-- }
```

Comment:
Why extension functions instead of members?

Reply:
Done